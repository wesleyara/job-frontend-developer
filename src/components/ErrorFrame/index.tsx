import { MdErrorOutline } from "react-icons/md";

interface ErrorFrameProps {
  error: Error | null;
}

export const ErrorFrame = ({ error }: ErrorFrameProps) => {
  return (
    <section
      data-cy="error-frame"
      className="mt-5 w-full px-3 flex flex-col items-center gap-3"
    >
      <MdErrorOutline size={50} />
      <span className="text-center">{error?.message}</span>
    </section>
  );
};
