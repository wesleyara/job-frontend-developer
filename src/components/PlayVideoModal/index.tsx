import { Dialog, DialogPanel, DialogTitle } from "@headlessui/react";
import { YoutubeDetailsType } from "~/types";
import { Dispatch, SetStateAction } from "react";

interface PlayVideoModalProps {
  item: YoutubeDetailsType;
  isOpen: boolean;
  setIsOpen: Dispatch<SetStateAction<boolean>>;
}

export const PlayVideoModal = ({
  item,
  isOpen,
  setIsOpen,
}: PlayVideoModalProps) => {
  return (
    <Dialog
      open={isOpen}
      onClose={() => setIsOpen(false)}
      className="relative z-50"
    >
      <div className="fixed inset-0 flex mx-auto items-center justify-center p-4">
        <DialogPanel className="space-y-4 border border-gray-300 bg-white p-12 max-w-[768px] w-[90%] text-center rounded-md relative">
          <button
            data-cy="close-youtube-modal"
            className="absolute top-2 right-3"
            onClick={() => setIsOpen(false)}
          >
            X
          </button>
          <DialogTitle className="text-lg">{item.snippet.title}</DialogTitle>

          <iframe
            data-cy="youtube-player"
            src={`https://www.youtube.com/embed/${item.id}`}
            title={item.snippet.title}
            className="inset-0 w-full h-[200px] sm:h-[400px]"
            allowFullScreen
          />
        </DialogPanel>
      </div>
    </Dialog>
  );
};
