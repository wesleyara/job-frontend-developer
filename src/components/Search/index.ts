import { SearchButton } from "./SearchButton";
import { SearchInput } from "./SearchInput";
import { SearchRoot } from "./SearchRoot";

export const Search = {
  Root: SearchRoot,
  Button: SearchButton,
  Input: SearchInput,
};
