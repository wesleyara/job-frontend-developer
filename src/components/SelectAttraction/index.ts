import { SelectAttractionButton } from "./SelectAttractionButton";
import { SelectAttractionContent } from "./SelectAttractionContent";
import { SelectAttractionRoot } from "./SelectAttractionRoot";

export const SelectAttraction = {
  Root: SelectAttractionRoot,
  Content: SelectAttractionContent,
  Button: SelectAttractionButton,
};
