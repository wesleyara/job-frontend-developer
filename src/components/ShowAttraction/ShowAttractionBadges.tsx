import { classificationArray } from "~/lib/utils";
import { Classification } from "~/types";

import { Badge } from "../ui/badge";

interface ShowAttractionBadgesProps {
  classifications: Classification[];
}

export const ShowAttractionBadges = ({
  classifications,
}: ShowAttractionBadgesProps) => {
  return (
    <span className="flex gap-2 justify-center flex-wrap mt-5 px-2">
      {classificationArray(classifications[0]).map(item => (
        <Badge key={item.name} variant="outline">
          {item.name}
        </Badge>
      ))}
    </span>
  );
};
