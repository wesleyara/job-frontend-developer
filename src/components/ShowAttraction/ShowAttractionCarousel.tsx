import { ImagesPerRatioTypes } from "~/types";
import Autoplay from "embla-carousel-autoplay";
import { useWidth } from "essentials-utils";
import Image from "next/image";
import { Dispatch, SetStateAction, useRef } from "react";

import {
  Carousel,
  CarouselContent,
  CarouselItem,
  CarouselNext,
  CarouselPrevious,
} from "../ui/carousel";

interface ShowAttractionCarouselProps {
  ratioSelected: string;
  imagesPerRatio: ImagesPerRatioTypes[];
  selected: ImagesPerRatioTypes | undefined;
  setRatioSelected: Dispatch<SetStateAction<string>>;
}

export const ShowAttractionCarousel = ({
  ratioSelected,
  imagesPerRatio,
  selected,
  setRatioSelected,
}: ShowAttractionCarouselProps) => {
  const plugin = useRef(Autoplay({ delay: 3000, stopOnInteraction: true }));

  const [width, height] = selected?.ratio?.split("_")?.map(Number) || [0, 0];

  const widthSize = useWidth();
  const multiplier = widthSize > 400 ? 80 : 60;
  const ajustedWidth =
    width >= 16 ? width * (multiplier / 4) : width * multiplier;
  const ajustedHeight =
    width >= 16 ? height * (multiplier / 4) : height * multiplier;

  return (
    <span className="flex flex-col items-center gap-4 mt-4 w-full relative">
      <h5>Coleção de imagens</h5>
      <select
        onChange={e => setRatioSelected(e.target.value)}
        value={ratioSelected}
        className="p-2 border border-gray-300 rounded-md bg-white"
      >
        <option value="">Selecione</option>
        {imagesPerRatio.map((item: any) => (
          <option key={item.ratio} value={item.ratio}>
            {item.title}
          </option>
        ))}
      </select>

      {ratioSelected && selected && (
        <span className="flex flex-col items-center">
          <Carousel
            plugins={[plugin.current]}
            className="sm:w-[500px] w-[300px]"
            onMouseEnter={plugin.current.stop}
            onMouseLeave={plugin.current.reset}
          >
            <CarouselContent>
              {selected.images.map(image => (
                <CarouselItem
                  key={image.url}
                  className="flex items-center justify-center"
                >
                  <Image
                    src={image.url}
                    alt={"Carousel image"}
                    width={ajustedWidth}
                    height={ajustedHeight}
                  />
                </CarouselItem>
              ))}
            </CarouselContent>
            {widthSize > 640 && (
              <>
                <CarouselPrevious />
                <CarouselNext />
              </>
            )}
          </Carousel>
        </span>
      )}
    </span>
  );
};
