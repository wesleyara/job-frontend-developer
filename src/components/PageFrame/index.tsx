"use client";

import { useData } from "~/context/DataContext";
import { useState } from "react";
import { FaSearch } from "react-icons/fa";

import { ErrorFrame } from "../ErrorFrame";
import { Search } from "../Search";
import { SelectAttraction } from "../SelectAttraction";
import { ShowAttraction } from "../ShowAttraction";

export const PageFrame = () => {
  const {
    searchQuery,
    ticketSearchData,
    selectedAttraction,
    ticketSearchIsLoading,
    ticketSearchError,
    handleChange,
    handleSelectAttraction,
  } = useData();

  const [ratioSelected, setRatioSelected] = useState<string>("");
  const [seeMore, setSeeMore] = useState<boolean>(false);

  const enableToRender = ticketSearchData && ticketSearchData.length > 0;

  const render = () => {
    if (ticketSearchIsLoading) {
      return <p className="mt-10">Carregando...</p>;
    }

    if (ticketSearchError) {
      return <ErrorFrame error={ticketSearchError} />;
    }

    if (selectedAttraction) {
      const imagesPerRatio: any = [];

      selectedAttraction.images.forEach(image => {
        const ratio = image.ratio;
        const find = imagesPerRatio.find((item: any) => item.ratio === ratio);

        if (find) {
          find.images.push(image);
        } else {
          imagesPerRatio.push({
            ratio,
            images: [image],
          });
        }
      });

      console.log(imagesPerRatio);

      const selected = imagesPerRatio.find(
        (item: any) => item.ratio === ratioSelected,
      );

      return (
        <ShowAttraction.Root onClick={() => setSeeMore(prev => !prev)}>
          {!seeMore ? (
            <>
              <ShowAttraction.Info selectedAttraction={selectedAttraction} />
              <ShowAttraction.Badges
                classifications={selectedAttraction.classifications}
              />
            </>
          ) : (
            <ShowAttraction.Carousel
              setRatioSelected={setRatioSelected}
              selected={selected}
              imagesPerRatio={imagesPerRatio}
              ratioSelected={ratioSelected}
            />
          )}
        </ShowAttraction.Root>
      );
    }

    if (enableToRender) {
      return (
        <div className="flex flex-col gap-3 mt-5 overflow-auto md:w-[600px] w-[90%]">
          {ticketSearchData.map(item => (
            <SelectAttraction.Root key={item.id}>
              <SelectAttraction.Content item={item} />
              <SelectAttraction.Button
                onClick={() => handleSelectAttraction(item)}
              />
            </SelectAttraction.Root>
          ))}
        </div>
      );
    }
  };

  return (
    <main className="flex flex-col items-center my-24">
      <Search.Root>
        <Search.Input onChange={handleChange} value={searchQuery} />
        <Search.Button icon={FaSearch} />
      </Search.Root>

      {render()}
    </main>
  );
};
