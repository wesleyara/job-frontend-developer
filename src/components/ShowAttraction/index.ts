import { ShowAttractionBadges } from "./ShowAttractionBadges";
import { ShowAttractionCarousel } from "./ShowAttractionCarousel";
import { ShowAttractionInfo } from "./ShowAttractionInfo";
import { ShowAttractionRoot } from "./ShowAttractionRoot";

export const ShowAttraction = {
  Root: ShowAttractionRoot,
  Info: ShowAttractionInfo,
  Badges: ShowAttractionBadges,
  Carousel: ShowAttractionCarousel,
};
