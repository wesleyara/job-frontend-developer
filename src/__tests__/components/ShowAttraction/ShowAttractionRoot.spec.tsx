import { render } from "@testing-library/react";
import { ShowAttraction } from "~/components/ShowAttraction";
import { beforeEach, describe, expect, it } from "vitest";

describe("SearchRoot unit test", () => {
  let sut: JSX.Element;

  beforeEach(() => {
    sut = (
      <ShowAttraction.Root>
        <span />
      </ShowAttraction.Root>
    );
  });

  it("should be defined", () => {
    expect(sut).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(sut);

    expect(container).toMatchSnapshot();
  });

  it("should have a section element", () => {
    const { container } = render(sut);

    expect(container.querySelector("section")).not.toBeNull();
  });

  it("should have a child element", () => {
    const { container } = render(sut);

    expect(container.querySelector("span")).not.toBeNull();
  });

  it("should have a tailwind class name", () => {
    const { container } = render(sut);

    const element = container.querySelector(".flex");

    expect(element).not.toBeNull();
  });
});
