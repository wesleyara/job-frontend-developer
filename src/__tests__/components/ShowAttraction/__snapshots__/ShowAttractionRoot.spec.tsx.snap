// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchRoot unit test > should render correctly 1`] = `
<div>
  <section
    class="flex flex-col items-center gap-5 mt-10 border-2 border-gray-300 rounded-lg bg-white py-10 w-[90%] max-w-[600px] relative"
  >
    <button
      class="absolute top-2 right-2 p-2 rounded-md"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          id="Circle_More"
        >
          <g>
            <circle
              cx="12.001"
              cy="12"
              r="1"
            />
            <circle
              cx="16.001"
              cy="12"
              r="1"
            />
            <circle
              cx="8.001"
              cy="12"
              r="1"
            />
            <path
              d="M12,21.932A9.934,9.934,0,1,1,21.934,12,9.944,9.944,0,0,1,12,21.932ZM12,3.065A8.934,8.934,0,1,0,20.934,12,8.944,8.944,0,0,0,12,3.065Z"
            />
          </g>
        </g>
      </svg>
    </button>
    <span />
  </section>
</div>
`;
